### ISSUE TITLE
possible bug in ssl certificate config

### ISSUE THREAD

**Issue by mcast** [irods/irods#4938]
##  Bug Report

This is a suggestion that something may be wrong with SSL certificate enforcement, but it would probably need more detail for you to act on it.

### iRODS Version, OS and Version

iRODS 4.2.7 on Ubuntu Bionic

### What did you try to do?

1. Enable and enforce SSL on the dev zone (done by @kript and @bh9)
    * noting that it is possible to enable SSL and then not enforce its use, which could lead to Surprises
2. Run icommands on dev zone, from our Openstack
3. Run Baton 2.0.1 on dev zone, from our Openstack
4. Run icommands on dev zone, from internal hosts
5. Run Baton 2.0.1 on dev zone, from internal hosts

### Expected behavior

Having found everything working on Openstack (2, 3), I expected that to continue on internal hosts with the same base OS (4, 5).

### Observed behavior (including steps to reproduce, if applicable)

At 5. Baton on internal hosts could not connect by SSL. At first I thought this was due to a build-time option because `ldd` has significantly different reports for the two builds; then my friendly sysads pointed to the missing `"irods_ssl_ca_certificate_file": "/etc/ssl/certs/ca-certificates.crt"` entry in the `irods_environment*.json`.

It isn't clear to me how Baton was finding the certificate file on Openstack and not the internal hosts. Various aspects of host configuration are handled differently.

My suspicion is that the icommands are failing to check the certificate, and it is for this reason they worked when incorrectly configured at 4. This is the possible bug.

**Comment by kript** [comment]
@mcast can I check that you saw this behaviour with SSL enabled in your `irods_environment.json` file (i.e. you weren't relying on the server enforcing it)? As we tried both with SSL enforced and not. Our deployment scenario is to have SSL enforced, so if the client doesn't also require it it will fail, AIUI.

aka `"irods_client_server_policy": "CS_NEG_REQUIRE",`

**Comment by mcast** [comment]
Here's a transcript with more detail,

```
mca@casm3-head2:~$ i-cgpdev ils /
/:
  C- /cgp_dev
mca@casm3-head2:~$ echo '{ "collection": "/cgp_dev/sandbox/" }' | i-cgpdev /software/sciops/pkgg/baton/2.0.1/bin/baton-list
{"collection": "/cgp_dev/sandbox"}
mca@casm3-head2:~$ perl -i~ -pe 's{irods_ssl_ca_certificate_file}{__$&}' ~/.irods/irods_environment-cgpdev.json
mca@casm3-head2:~$ i-cgpdev ils /
/:
  C- /cgp_dev
mca@casm3-head2:~$ echo '{ "collection": "/cgp_dev/sandbox/" }' | i-cgpdev /software/sciops/pkgg/baton/2.0.1/bin/baton-list
ERROR: [-]      iRODS/lib/core/src/sockComm.cpp:911:connectToRhost :  status [SSL_HANDSHAKE_ERROR]  errno [] -- message []
        [-]     iRODS/lib/core/src/sockComm.cpp:67:sockClientStart :  status [SSL_HANDSHAKE_ERROR]  errno [] -- message [failed to call 'client start']
                [-]     libssl.cpp:695:ssl_client_start :  status [SSL_HANDSHAKE_ERROR]  errno [] -- message [error in SSL_connect | error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed]

ERROR: _rcConnect: connectToRhost error, server on irods-cgp-dev-ies1.internal.sanger.ac.uk:1247 is probably down status = -2103000 SSL_HANDSHAKE_ERROR
2020-05-14T10:45:43Z ERROR Failed to connect to irods-cgp-dev-ies1.internal.sanger.ac.uk:1247 zone 'cgp_dev' as 'mca'
mca@casm3-head2:~$ echo $?
0
```

That last `$?==0` looks like an aspect of https://github.com/wtsi-npg/baton/issues/218

and the config file

```
mca@casm3-head2:~$ diff -U100 .irods/irods_environment-cgpdev.json{~,}
--- .irods/irods_environment-cgpdev.json~       2020-05-14 11:43:50.730283000 +0100
+++ .irods/irods_environment-cgpdev.json        2020-05-14 11:45:31.074003000 +0100
@@ -1,21 +1,21 @@
 {
     "irods_zone_name": "cgp_dev",
     "irods_host": "irods-cgp-dev-ies1.internal.sanger.ac.uk",
     "irods_port": 1247,
 
     "irods_default_resource": "green",
 
 
     "irods_encryption_key_size": 32,
     "irods_encryption_salt_size": 8,
     "irods_encryption_num_hash_rounds": 16,
     "irods_encryption_algorithm": "AES-256-CBC",
     "irods_client_server_negotiation": "request_server_negotiation",
     "irods_client_server_policy": "CS_NEG_REQUIRE",
-    "irods_ssl_ca_certificate_file": "/etc/ssl/certs/ca-certificates.crt",
+    "__irods_ssl_ca_certificate_file": "/etc/ssl/certs/ca-certificates.crt",
 
     "irods_user_name": "mca",
     "irods_authentication_file": "/nfs/users/nfs_m/mca/.irods/.irodsA-cgpdev",
     "irods_home": "/cgp_dev/home/mca",
     "irods_cwd": "/cgp_dev/home/mca"
 }
```

Aside: I have found it's easier to get config files straight if I order the keys, so that's why the whitespace is odd.

**Comment by mcast** [comment]
Extra information (but still incomplete, there are more experiments that could be done)

I realised that the Baton on Openstack (2), (3) above were with icommands 4.2.7 and associated libraries. With another host also on Openstack but with icommands-4.1.12

## What did you try to do?

With the dev zone continuing to be 4.2.7 with mandatory SSL,

6. Run icommands 4.1.12 on dev zone, from internal hosts
7. Run Baton 2.0.1 on dev zone, from internal hosts

I get the same pattern: icommands work, Baton doesn't, until I mend the configuration file.

**Comment by trel** [comment]
Can you say a bit more about how you think we can test this?

> My suspicion is that the icommands are failing to check the certificate, and it is for this reason they worked when incorrectly configured at 4. This is the possible bug.

We would...
a) configure a server as `CS_NEG_REQUIRE`
b) configure the iCommands on a different server as `CS_NEG_DONT_CARE` or `CS_NEG_REQUIRE`
c) misconfigure the iCommands' SSL cert

Expect to see the iCommands work?  Because they're ignoring the misconfiguration in c) ?

What is special about the hostnames / internal networking / machines that are running the iRODS server and the iCommands?

**Comment by bh9** [comment]
Is there a possibility that the icommands are looking at the known root CAs (on ubuntu that is `/etc/ssl/certs/ca-certificates.crt`) which would make this not a bug in icommands, but a feature which could potentially be added into baton?

**Comment by trel** [comment]
Hmmm, very interesting.

iRODS does not keep its own keystore (like Java, etc.), so... perhaps @bh9 is right?

**Comment by trel** [comment]
I expect this has been resolved - there was a lot of work in the run-up to iRODS 5.0 on the TLS/SSL codepaths.

**Comment by korydraughn** [comment]
> Is there a possibility that the icommands are looking at the known root CAs (on ubuntu that is `/etc/ssl/certs/ca-certificates.crt`) which would make this not a bug in icommands, but a feature which could potentially be added into baton?

This is true. It has been reported that the icommands do not require specifying TLS configuration in the irods_environment.json file.

See https://groups.google.com/d/msgid/irod-chat/14fd5be2-0ee8-4fc9-b9cb-5b7f7cb30021n%40googlegroups.com?utm_medium=email&utm_source=footer.
